### PR TITLE
Otimização - Adiciona code-splitting por página

### LINKED ISSUE #1401: ajustar bundle da pilot para tela de login
### Contexto
Atualmente temos apenas dois bundles na pilot, um para vendors e outro para os demais pacotes. Devido a isso o load inicial acaba baixando mais coisas que precisa no momento.
A ideia é criar um bundle especifico para tela de login para conseguir reduzir a carga da pagina de login.

### Descrição técnica
Ajustar os bundles da pilot para cria rum bundle especifico para a tela de login e possibilitar o carregamento assíncrono dos demais bundles conforme a necessidade das rotas da aplicação.

### Requisitos e Critérios de Aceite
- [ ] criar bundle para pagina de login
- [ ] ajusta a aplicação para o carregamento assincrono dos bundles necessários nas demais rotas

### PR DESCRIPTION
**Pull request by vagnervst**
## Contexto
Precisamos quebrar o bundle principal da Pilot em mais de um arquivo para que o usuário carregue somente o necessário, no momento certo. Com isto, a velocidade de carregamento da aplicação diminuirá significativamente.

## Checklist
- [x] Separa as páginas de Account em diferentes `chunk`
- [x] Separa as páginas de LoggedArea em diferentes `chunk`
- [x] Exibe um componente de loading enquanto os assets da página são carregados

## Issues linkadas
- [x] Resolves #1401

## Screenshots
image

## Notas de deploy
Agora que splitamos o bundle em diversos chunks, o build está consumindo mais memória. Em minha máquina tive que permitir que o node consumisse um pouco mais (2GB), rodando o build da seguinte forma:
```bash
NODE_OPTIONS=--max-old-space-size=2000 yarn build
```

## Como testar?
Rode a pilot e navegue por todas as páginas, para cada uma delas deve ser carregado um `chunk.js` com seu respectivo nome.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 40a44817f818211a98a0739aa75b1bec6517ddc8
  * lways use the standard `scope: action description`, sempre em inglês  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 40a44817f818211a98a0739aa75b1bec6517ddc8
  * lways use the standard `scope: action description`, sempre em inglês  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 40a44817f818211a98a0739aa75b1bec6517ddc8
  * lways use the standard `scope: action description`, sempre em inglês  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 40a44817f818211a98a0739aa75b1bec6517ddc8
  * lways use the standard `scope: action description`, sempre em inglês  
    - The **scope** shouldn't have spaces. If the word is composite, use dashes.  
    - The **action** should be **add**, **remove**, **refactor**, **move**,  **bump**, **update** or **release**  
    - The **description** should contain a quick description of the change

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: c692f1372c2363a456fdad7caa7bb2ce60f2c65d
  * Don't use more than 72 characters per line in the body of your commit. If you reach this limit at the same line, just break and continue at next line. To separate paragraphs, leave a blank line between them.

The complete commit style guide is available at https://github.com/pagarme/git-style-guide

---
This message was auto-generated by https://gitcop.com

File: packages/pilot/src/pages/LoggedArea/routes.js [hunk 1]
```diff
@@ -1,18 +1,42 @@
...
+  /* webpackChunkName: "company-settings" */ '../CompanySettings'
+))
+
+const Withdraw = lazy(() => import(
+  /* webpackChunkName: "withdraw" */ '../Withdraw'
+))
+
+const Recipients = lazy(() => import(
+  /* webpackChunkName: "recipients" */ '../Recipients'
+))
```
Comment (miglsoares) [inline packages/pilot/src/pages/LoggedArea/routes.js]: Aqui também acho interessante colocar num arquivo separado nomeado `webpackChunkImports.js` e importar apenas as funções.

File: packages/pilot/src/pages/Transactions/index.js [hunk 2]
```diff
@@ -1,11 +1,19 @@
...
+import details from './Details'
+import search from './Search'
+
+const TransactionDetails = lazy(() => import(
+  /* webpackChunkName: "transaction-details" */ './Details/Details'
+))
+
+const TransactionsSearch = lazy(() => import(
+  /* webpackChunkName: "transactions" */ './Search/Search'
+))
```
Comment (miglsoares) [inline packages/pilot/src/pages/Transactions/index.js]: Aqui também acho interessante colocar num arquivo separado nomeado webpackChunkImports.js e importar apenas as funções.

File: packages/pilot/src/pages/Recipients/index.js [hunk 3]
```diff
@@ -1,12 +1,22 @@
...
+  /* webpackChunkName: "recipients" */ './Search/Search'
+))
+
+const RecipientDetail = lazy(() => import(
+  /* webpackChunkName: "recipient-detail" */ './Detail'
+))
+
+const RecipientsAdd = lazy(() => import(
+  /* webpackChunkName: "recipient-add" */ './Add'
+))
```
Comment (miglsoares) [inline packages/pilot/src/pages/Recipients/index.js]: Aqui também acho interessante colocar num arquivo separado nomeado webpackChunkImports.js e importar apenas as funções.

File: packages/pilot/src/pages/Account/index.js [hunk 4]
```diff
@@ -15,21 +15,49 @@ import {
...
+))
+
+const UserSignUp = lazy(() => import(
+  /* webpackChunkName: "user-signup" */ './SignUp/User'
+))
+
+const UserSignupConfirmation = lazy(() => import(
+  /* webpackChunkName: "user-signup-confirmation" */ './SignUp/User/Confirmation'
+))
+
```
[the remark this answers is not in the document]
Comment (vagnervst) [reply packages/pilot/src/pages/Account/index.js]: Boa, feito!

**Comment by vagnervst** [comment]
@miglsoares Isso ocorre pois o download do chunk é rápido demais, assim o loading aparece e some muito rápido, vou investigar se há uma forma de melhorar isto

**Comment by rsmelo** [comment]
@vagnervst o @Deivis publicou a versão nova do react-scripts pode bumpar ele no pr https://github.com/pagarme/react-scripts-former-kit-dashboard/releases/tag/2.1.1

**Comment by vagnervst** [comment]
> Não sei se é o comportamento esperado, mas quando troco de página sem ter carregado o `chunk` da página que preciso ir, ele pisca um pouco, não mostra nenhum loading. Vi pela descrição do PR que existia um loading, então não sei se isso era esperado de fato.

Corrigido @miglsoares, ajustei os Loader para aparecerem em uma área menor e sem background.

File: .gitignore [hunk 5]
```diff
@@ -59,5 +59,7 @@ typings/
 
 # distribution files
 dist
+build
```
Comment (Deivis) [inline .gitignore]: Por que a pasta `build` foi adicionada ao `.ignore` ?

Comment (rsmelo) [reply .gitignore]: acho que por que ela nunca é commitada mesmo

Comment (vagnervst) [reply .gitignore]: Tive que gerar os builds algumas vezes para testar, e ao tentar fazer o checkout para outra branch para fazer algum review, o git reclama dizendo que tenho mudanças não comittadas. Isso atrapalha um pouco então achei uma boa adicionar estas pastas no `.gitignore`

Exemplo ao remover a condição:
image

File: packages/pilot/src/pages/LoggedArea/index.js [hunk 6]
```diff
@@ -14,10 +14,13 @@ import {
   path,
 } from 'ramda'
 import { Layout } from 'former-kit'
-import env from '../../environment'
+
 import { ErrorBoundary } from '../ErrorBoundary'
 import Sidebar from './Sidebar'
 import Header from './Header'
+import Loader from '../../components/Loader'
+
```
Comment (Deivis) [inline packages/pilot/src/pages/LoggedArea/index.js]: Não precisa desta linha em branco.

File: packages/pilot/src/pages/LoggedArea/index.js [hunk 7]
```diff
@@ -14,10 +14,13 @@ import {
   path,
 } from 'ramda'
 import { Layout } from 'former-kit'
-import env from '../../environment'
+
```
Comment (Deivis) [inline packages/pilot/src/pages/LoggedArea/index.js]: Não precisa desta linha em branco.

File: packages/pilot/src/pages/LoggedArea/index.js [hunk 8]
```diff
@@ -86,16 +89,26 @@ const LoggedArea = ({
...
-        {Object.values(routes).map(({ component, path: pathURI }) => (
-          <Route
-            key={pathURI}
-            path={pathURI}
-            component={component}
+      <Suspense
+        fallback={(
+          <Loader
+            text={t('loading')}
+            position="relative"
```
Comment (marisamodolo) [inline packages/pilot/src/pages/LoggedArea/index.js]: No console aparece esse warning depois do login:
image

File: packages/pilot/.storybook/config.js [hunk 9]
```diff
@@ -1,9 +1,9 @@
 import './themeDecorator'
 import './percyIntegration'
-import { setOptions } from '@storybook/addon-options'
+import { withOptions } from '@storybook/addon-options'
 
-setOptions({
-  name: 'Pilot',
-  url: 'https://github.com/pagarme/pilot',
+withOptions({
```
Comment (rsmelo) [inline packages/pilot/.storybook/config.js]: withOptions é o novo meio de fazer o que setOptions fazia na versão anterior?

File: packages/pilot/src/pages/LoggedArea/index.js (hunk 6 above)
Comment (vagnervst) [reply packages/pilot/src/pages/LoggedArea/index.js]: Corrigido!

File: packages/pilot/src/pages/LoggedArea/index.js (hunk 7 above)
Comment (vagnervst) [reply packages/pilot/src/pages/LoggedArea/index.js]: Deixei esta linha em branco para separar os imports de packages e dos nossos modules internos @Deivis

File: packages/pilot/src/pages/LoggedArea/index.js (hunk 8 above)
Comment (vagnervst) [reply packages/pilot/src/pages/LoggedArea/index.js]: Boa, corrigido!

File: packages/pilot/.storybook/config.js (hunk 9 above)
Comment (vagnervst) [reply packages/pilot/.storybook/config.js]: Exato! o `setOptions` foi deprecado e seu uso passou a exibir warnings no console

**Pull request merged**